src/string_id.h

```diff
@@ -69,69 +107,96 @@ template<typename T>
 class string_id
 {
     public:
+        template<class U>
+        friend class string_id;
         using This = string_id<T>;
 
-        /**
-         * Forwarding constructor, forwards any parameter to the std::string
-         * constructor to create the id string. This allows plain C-strings,
-         * and std::strings to be used.
-         */
-        // Beautiful C++11: enable_if makes sure that S is always something that can be used to constructor
-        // a std::string, otherwise a "no matching function to call..." error is generated.
+        template<unsigned N>
+        explicit constexpr string_id( const char ( &str )[N] ) : _hash( Hash::CompileTime::fnv1_64( str ) ),
+            _char_ptr( str ), _is_char_ptr( true ) {}
+
+        template<typename U>
+        string_id( const string_id<U> &other ) : _hash( other.hash() ), _char_ptr( other._char_ptr ),
+            _is_char_ptr( other._is_char_ptr ) {}
+
         template<typename S, class = typename
-                 std::enable_if< std::is_convertible<S, std::string >::value>::type >
-        explicit string_id( S &&
-                            id ) : _id( std::forward<S>( id ) ), _cid( INVALID_CID ), _version( INVALID_VERSION ) {}
+                 std::enable_if< std::is_convertible< S, std::string >::value >::type
+                 >
+        explicit string_id( S && id ) {
+            auto str = std::make_unique<std::string>( std::forward<S>( id ) );
+            uint64_t hash = Hash::RunTime::fnv1_64( str->c_str(), str->size() );
+            _hash = hash;
+
+            auto insertion = cached_strings().try_emplace( hash, std::move( str ) );
+            auto cached_string = insertion.first->second.get();
+
+            // Verify we don't have hash collisions
+            //assert(insertion.second || *cached_string == id);
+
+            _string = cached_string;
+        }
+
         /**
          * Default constructor constructs an empty id string.
          * Note that this id class does not enforce empty id strings (or any specific string at all)
          * to be special. Every string (including the empty one) may be a valid id.
          */
-        string_id() : _cid( INVALID_CID ), _version( INVALID_VERSION ) {}
+        constexpr string_id() : string_id( "" ) {};
         /**
          * Comparison, only useful when the id is used in std::map or std::set as key. Compares
          * the string id as with the strings comparison.
          */
         bool operator<( const This &rhs ) const {
-            return _id < rhs._id;
+            // Not str() compares to prevent needless tier ups to std::string
+            return strcmp( c_str(), rhs.c_str() ) < 0;
```

Comment:
Currently the biggest slowdown in string ids comes from comparison, as it's used in `std::set` and `cata::flat_set`. To get a measurable performance change, all of comparison, equality and hash must be sped up.

Reply:
I haven't dug into this, but when I replaced this with hash comparisons, many things broke, so I assumed someone did something foul and assumed it was a string...
I'll check into this

Reply:
Please do. There should be no reason for things to break if hash works as expected. That might be an indication of a deeper problem.